src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs

```diff
@@ -96,71 +105,176 @@ public ImmutableArray<Location>.Builder UsageLocations
+            if (!TryGetDictionaryTypeAndMembers(context.Compilation, out var iDictionaryType, out var containsKeySymbol, out var addSymbol))
             {
                 return;
             }
 
-            compilationContext.RegisterOperationAction(context =>
-                OnOperationAction(context, iDictionaryType, containsKeySymbol), OperationKind.Invocation);
+            context.RegisterOperationAction(ctx => OnInvocationOperation(iDictionaryType, containsKeySymbol, addSymbol, ctx), OperationKind.Invocation);
         }
 
-        private static void OnOperationAction(OperationAnalysisContext context, INamedTypeSymbol dictionaryType,
-            IMethodSymbol containsKeySymbol)
+        private static void OnInvocationOperation(INamedTypeSymbol iDictionaryType, IMethodSymbol containsKeySymbol, IMethodSymbol addSymbol, OperationAnalysisContext context)
         {
             var containsOperation = (IInvocationOperation)context.Operation;
-            if (!IsContainsKeyMethod(containsOperation.TargetMethod, containsKeySymbol) ||
-                !IsDictionaryType(containsOperation.GetReceiverType(context.Compilation, true,
-                    context.CancellationToken), dictionaryType))
+            if (!IsContainsKeyMethod(containsOperation.TargetMethod, containsKeySymbol))
             {
                 return;
             }
 
-            var usageContext = new DictionaryUsageContext(containsOperation.Instance, containsOperation.Arguments[0].Value);
-            if (!GetParentConditionalOperation(containsOperation, ref usageContext, out var conditionalOperation, out var guardsTruePath))
+            var suspectedDictionaryType = containsOperation.GetReceiverType(context.Compilation, true, context.CancellationToken);
+            if (!IsDictionaryType(suspectedDictionaryType, iDictionaryType))
+            {
                 return;
+            }
 
-            var guardedPath = guardsTruePath ? conditionalOperation.WhenTrue : conditionalOperation.WhenFalse;
-            if (guardedPath != null)
-                FindUsages(guardedPath, ref usageContext);
-            else if (!guardsTruePath && HasReturnOrSetsKeyInTruePath(conditionalOperation, ref usageContext))
-                FindUsageInOperationsAfterConditionBlock(conditionalOperation, ref usageContext);
+            ReportGuardedDictionaryPattern(SearchContext.Indexer, PreferTryGetValueDiagnostic);
+            ReportGuardedDictionaryPattern(SearchContext.AddMethod, PreferTryAddDiagnostic);
 
-            if (usageContext.UsageLocations.Count == 0)
-                return;
+            void ReportGuardedDictionaryPattern(SearchContext searchContext, DiagnosticDescriptor diagnosticDescriptor)
+            {
+                if (searchContext == SearchContext.AddMethod && suspectedDictionaryType.GetMembers(TryAdd).IsEmpty)
+                {
+                    return;
+                }
```

Comment:
Does this check need to be here? Could we move this into `else if (searchContext == SearchContext.AddMethod)` conditional block below?

Reply:
I wanted to exit as early as possible if the dictionary does not have a `TryAdd` method. But I can still move it, if you would like me to.

Reply:
Ok, you can keep it